### ISSUE TITLE
`stream_socket_client()` fails to connect to certain sites when ignoring SSL certificate check

### ISSUE THREAD

**Issue by yuliu** [mybb/mybb#3906]
Relevant codes are:
https://github.com/mybb/mybb/blob/mybb_1822/inc/functions.php#L7172-L7180

In this part, MyBB tries to connect to a HTTPS site (hostname `example.com` and **IP `a.b.c.d`** and port `XXX`) using `ssl://a.b.c.d:XXX`, **and with ([SSL context options](https://www.php.net/manual/en/context.ssl.php)) only `verify_peer` & `verify_peer_name` set to `false` to turn off SSL verification & peer name, or CN, verification.**

However, for certain sites that don't provide any certificate via direct IP connection, above connection will fail, probably with following PHP warning:
```
Warning [2] stream_socket_client(): SSL operation failed with code 1. OpenSSL Error messages: error:14094410:SSL routines:ssl3_read_bytes:sslv3 alert handshake failure
```

The reason here is that no `peer_name` provided for such connection in `stream_socket_client()`, and for certain sites, the connection will be terminated from remote server with failed handshake.

It's so interesting that I find this issue during finding why & how a user's MyBB installation can't check version. I'll explain a bit more in a following post.

**Comment by yuliu** [comment]
TL;DR:
- mybb.com at Cloudflare changed something (SSL certificates/IPs), breaking MyBB forum's connection to mybb.com.
- The real bug is in MyBB where an option is missing for connecting to a https:// (SSL/TLS) site, resulting in connection failure in a particular circumstance.

OK, long story when troubleshooting a user's forum can't get MyBB version check:

- Since MyBB uses any one of `curl`, `stream_socket_client` and `fsockopen` to connect to a web site, the problem could be caused by many reasons: PHP modules/extensions not installed, wrong configurations, OpenSSL issues (for HTTPS, SSL, TLS, etc. connections), PHP's disabled functions, PHP directives disallowing certain operations (`allow_url_open`, etc.). BTW, I think for some functions in MyBB, we should provide some discovery tools for investigations.

- Tried and tried and tested and tested, none of above is really causing the problem. It was once a while in my mind it's the OpenSSL certificates issue, but turned out it wasn't.

- **To reproduce the issue:**
  - make sure `php-curl` extension is disabled,
  - don't put any setting for directives `openssl.cafile` & `openssl.capath` in PHP.ini (or relevant PHP override settings),
  - check version at MyBB AdminCP: it probably will fail (at least now, but will success days ago, continue reading).

- Then the problem was tracked down into this part of code: https://github.com/mybb/mybb/blob/mybb_1822/inc/functions.php#L7172-L7180 . The reason here is that no `peer_name` provided for such connection in `stream_socket_client()`, and for certain sites (like new mybb.com proxied by Cloudflare), the connection will be terminated from remote server with failed handshake.


The user who had this issue was able to check version days ago, but without any changes on his server, now the version check failed. That's why the problem is starting to flow recently.

I happened to check the certificate at https://mybb.com/ which was issued on 3/19, some days ago. This certificate is different from previous one, [details here](https://community.cloudflare.com/t/cloudflare-universal-and-universal-shared-certificates/59523). And:

- Before the new certificate, mybb.com proxied by Cloudflare, together with some other web sites, were allocated with a shared IP (maybe `104.27.0.106`). A Universal SSL certificate is associated with this shared IP which has mybb.com and other web sites domain names in SAN. That's the legacy scenario used by Cloudflare, and when you visiting https://104.27.0.106 in a web browser, it'll prompt you with a "Security Risk" warning saying that the hostname (actually it's an IP as hostname) is not listed in the associated certificate. **From here we know, connecting this previous allocated IP by TLS, the server does response.**

- With the new certificate, mybb.com is re-allocated with another (probably shared, I'm not sure) IP (seems `104.24.30.89`), and this IP doesn't get associated with any certificate. This means that if you visit https://104.24.30.89 in a browser, Firefox will probably notify with error "SSL_ERROR_NO_CYPHER_OVERLAP". **It's not so certain that direct IP access will result in failure, but we could try connecting using openssl.**

OK, using openssl to connect with IP address:
- For the old IP: `openssl s_client -connect 104.27.0.106:443` will connect and send you some certificate information (_SSL handshake is succesful_) and wait for input. Here you may use `GET / HTTP/1.0` and `Host: mybb.com` and `Connection: Close` to get mybb.com's page.

-For the new IP: `openssl s_client -connect 104.24.30.89:443` will try to connect but fails with `no peer certificate available`:
```CONNECTED(00000005)
140060371096000:error:14094410:SSL routines:ssl3_read_bytes:sslv3 alert handshake failure:../ssl/record/rec_layer_s3.c:1528:SSL alert number 40
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 7 bytes and written 314 bytes
Verification: OK
---
New, (NONE), Cipher is (NONE)
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 0 (ok)
---
```

- Still for the new IP, but **with a server name**: **`openssl s_client -connect 104.24.30.89:443 -servername mybb.com`** will process normally, connect successfully and send you the certificate information and wait for your input.

Now we know we need to pass the real hostname/servername when making a SSL/TLS connection using `stream_socket_client()`. If not doing so, the TLS handshake will fail, regardless of whether the SSL certificate verification is used.

**Comment by euantorano** [comment]
Thanks so much for troubleshooting this - seems like a real rabbit hole that you’ve been down!
